### PR TITLE
feat: implementation of overload reset method in swift classes

### PR DESCRIPTION
**Pull request by SKannaniOS**
## Description
This PR implements an overloaded `reset()` method in the Swift Analytics SDK that allows selective resetting of user identity data. The enhancement introduces new `ResetOptions` and `ResetEntries` classes that provide granular control over which user data components are reset, including anonymous ID, user ID, traits, and session information.

The implementation addresses the need for more flexible reset functionality, allowing developers to reset only specific parts of user identity data instead of performing a complete reset operation. This is particularly useful in scenarios where certain user data should be preserved while other components need to be cleared.

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactor/optimization

## Implementation Details
- **New `ResetOptions` class**: Configuration class for customizing user data reset behavior, encapsulating reset preferences through `ResetEntries`
- **New `ResetEntries` class**: Provides granular control over specific reset operations with boolean flags for `anonymousId`, `userId`, `traits`, and `session` (all default to `true`)
- **Enhanced `reset()` method**: Added overloaded version `reset(options: ResetOptions = ResetOptions())` with backward compatibility
- **Updated `ResetUserIdentityAction`**: Modified to accept `ResetEntries` parameter and perform conditional resets based on specified options
- **Enhanced `UserIdentity` extension**: Updated `resetUserIdentity` method with individual component reset methods (`resetUserId`, `resetTraits`) for granular control
- **Comprehensive test coverage**: Added 400+ lines of unit tests covering all reset scenarios, including selective resets, edge cases, and storage consistency
- **Improved test infrastructure**: Enhanced async test helpers and converted test suite to Swift Testing framework

### Key Files Modified:
- `Analytics.swift`: Added overloaded `reset(options:)` method with conditional session refresh
- `ResetOptions.swift`: New configuration class with comprehensive documentation
- `ResetEntries.swift`: New granular control class with individual reset flags
- `ResetUserIdentityAction.swift`: Enhanced to support selective reset operations based on entries
- `UserIdentity.swift`: Added granular reset methods and updated main reset logic
- `MockProvider.swift`: Improved test helpers with async support and better Given-When-Then pattern
- `RudderStackAnalyticsTests.swift`: Added comprehensive reset behavior test suite with 10 new test cases
- `ResetUserIdentityActionTests.swift`: Updated to use Swift Testing framework and added selective reset tests

## Checklist
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [x] I have checked for potential performance impacts and optimized if necessary.
- [x] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

### Sample Code Usage:

```swift
import RudderStackAnalytics

// Initialize analytics client
let analytics = Analytics(configuration: configuration)

// Identify a user with traits
analytics.identify(userId: "user123", traits: ["name": "John Doe", "email": "john@example.com"])

// Default reset behavior (resets all user data - backward compatible)
analytics.reset()

// Selective reset examples:

// Reset only anonymous ID, preserve user ID, traits, and session
let resetAnonymousIdOnly = ResetOptions(entries: ResetEntries(
    anonymousId: true,
    userId: false,
    traits: false,
    session: false
))
analytics.reset(options: resetAnonymousIdOnly)

// Reset only user ID and traits, preserve anonymous ID and session
let resetUserData = ResetOptions(entries: ResetEntries(
    anonymousId: false,
    userId: true,
    traits: true,
    session: false
))
analytics.reset(options: resetUserData)

// Reset everything except traits
let preserveTraits = ResetOptions(entries: ResetEntries(
    anonymousId: true,
    userId: true,
    traits: false,
    session: true
))
analytics.reset(options: preserveTraits)

// Session-only reset (useful for user switching scenarios)
let sessionReset = ResetOptions(entries: ResetEntries(
    anonymousId: false,
    userId: false,
    traits: false,
    session: true
))
analytics.reset(options: sessionReset)
```

### Automated Testing:
Run the comprehensive test suite that includes:
- **Default reset behavior verification** (backward compatibility)
- **Individual component resets** (anonymous ID, user ID, traits, session)
- **Multiple component combinations**
- **Edge cases** (no options, after shutdown, without previous identity)
- **Storage consistency verification**

### Manual Testing Scenarios:
1. **Backward Compatibility**: Call `analytics.reset()` and verify all user data is cleared
2. **Selective Anonymous ID Reset**: Test anonymous ID regeneration while preserving other data
3. **User Switching**: Test selective user ID reset for user switching scenarios
4. **Privacy Compliance**: Test selective traits reset for privacy compliance scenarios
5. **Session Management**: Test session-only reset for session management scenarios

## Breaking Changes
**None** - This is a non-breaking change. The existing `reset()` method continues to work exactly as before, with the new overloaded method providing additional functionality through an optional parameter with sensible defaults.

## Maintainers Checklist
- [ ] The code has been reviewed.
- [ ] CI tests have passed.
- [ ] All necessary documentation has been updated.

## Screenshots (if applicable)
Not applicable - this is a backend SDK enhancement with no UI components.

## Additional Context
- **Architecture**: The implementation follows the existing state management pattern using actions and reducers
- **Performance**: Minimal impact as selective reset logic uses simple boolean checks
- **Testing Strategy**: Comprehensive test coverage includes both unit tests for individual components and integration tests for end-to-end scenarios
- **Documentation**: All new public APIs include comprehensive documentation with parameter descriptions and usage examples
- **Future Compatibility**: The design allows for easy extension of additional reset options without breaking changes
- **Migration Path**: No migration required - existing code continues to work unchanged while new functionality is available through the overloaded method

### CODE REVIEW

File: Tests/RudderStackAnalyticsTests/RudderStackAnalyticsTests.swift [hunk 1]
```diff
@@ -350,3 +355,404 @@ extension RudderStackAnalyticsTests {
...
+extension RudderStackAnalyticsTests {
+    
+    func test_reset_defaultOptions_resetsAllUserIdentityData() {
+        guard let client = analytics_disk else { return XCTFail("No disk client") }
+        
+        given("Analytics client with identified user, traits, and active session") {
+            // Set up user identity
+            client.identify(userId: "test_user_123", traits: ["name": "John Doe", "email": "john@example.com"])
+            
+            runAfter(0.3) {
```
[the remark this answers is not in the document]
Comment (SKannaniOS) [reply Tests/RudderStackAnalyticsTests/RudderStackAnalyticsTests.swift]: No need to change it.

File: Sources/RudderStackAnalytics/StateManagement/ResetOptions.swift [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    public init(entries: ResetEntries = ResetEntries()) {
+        self.entries = entries
+    }
+}
+
+// MARK: - ResetEntries
+/**
+ Class representing specific entries to reset during a user data reset operation.
+ */
+public class ResetEntries {
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/StateManagement/ResetOptions.swift]: Let's move this to a new file. In future, there could be multiple other classes similar to `ResetEntries`, if those things are added over here, it'll bloat the class. Since the file name specifies `ResetOptions` let's have only that class.

File: Sources/RudderStackAnalytics/Base/Analytics.swift [hunk 3]
```diff
@@ -208,15 +208,17 @@ extension Analytics {
...
+     Resets the user identity information, using the provided options to determine which data to reset.
      */
-    public func reset() {
+    public func reset(options: ResetOptions = ResetOptions()) {
         guard self.isAnalyticsActive else { return }
         
-        self.userIdentityState.dispatch(action: ResetUserIdentityAction())
-        self.userIdentityState.state.value.resetUserIdentity(storage: self.storage)
+        self.userIdentityState.dispatch(action: ResetUserIdentityAction(options: options))
+        self.userIdentityState.state.value.resetUserIdentity(storage: self.storage, options: options)
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/Base/Analytics.swift]: Let us pass only what's required by those actions and nothing more (this is the best practice). In this case, we can simply pass `options.entries`.

File: Sources/RudderStackAnalytics/Base/Analytics.swift [hunk 4]
```diff
@@ -208,15 +208,17 @@ extension Analytics {
     }
     
     /**
-     Resets the user identity state by clearing stored identifiers and traits.
+     Resets the user identity information, using the provided options to determine which data to reset.
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/Base/Analytics.swift:211]: Please refer to the Kotlin PR and update the API documentation to indicate that the RESET API’s optional parameter is intended to `override` the default behaviour. The documentation should clearly describe both the behaviours: `default` and `options` behaviour.

Also, add a `- Parameter option: ...` in the API doc.

File: Sources/RudderStackAnalytics/StateManagement/ResetOptions.swift [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+//  RudderStackAnalytics
+//
+//  Created by Satheesh Kannan on 08/09/25.
+//
+
+import Foundation
+
+// MARK: - ResetOptions
+/**
+ Class representing options for resetting user data in the analytics SDK.
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/StateManagement/ResetOptions.swift]: Please refer to the Kotlin PR, or [here](https://github.com/rudderlabs/rudder-sdk-kotlin/blob/3c8e230a749b376bcff3952f448096918877a23b/core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/models/reset/ResetOptions.kt) and update the doc accordingly

File: Sources/RudderStackAnalytics/StateManagement/ResetOptions.swift [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+       - entries: An instance of `ResetEntries` specifying which data to reset. Default is a new instance with all entries set to `true`.
+     */
+    public init(entries: ResetEntries = ResetEntries()) {
+        self.entries = entries
+    }
+}
+
+// MARK: - ResetEntries
+/**
+ Class representing specific entries to reset during a user data reset operation.
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/StateManagement/ResetOptions.swift]: Please refer to the Kotlin PR, or [here](https://github.com/rudderlabs/rudder-sdk-kotlin/blob/3c8e230a749b376bcff3952f448096918877a23b/core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/models/reset/ResetEntries.kt) and update the doc accordingly.

File: Tests/RudderStackAnalyticsTests/RudderStackAnalyticsTests.swift [hunk 7]
```diff
@@ -75,13 +75,18 @@ extension RudderStackAnalyticsTests {
     func test_screenEvent_disk() async {
         guard let client = analytics_disk else { return XCTFail("No disk client") }
         client.screen(screenName: "Screen Event", category: "Main", properties: ["prop": "value"])
-        try? await Task.sleep(nanoseconds: 300_000_000)
-        await client.configuration.storage.rollover ()
+        try? await Task.sleep(nanoseconds: 500_000_000) // Increased delay
```
Comment (1abhishekpandey) [inline Tests/RudderStackAnalyticsTests/RudderStackAnalyticsTests.swift]: Let's add a TODO to indicate this sleep statement needs to be removed.

File: Tests/RudderStackAnalyticsTests/StateManagementTests/ResetUserIdentityActionTests.swift
Comment (1abhishekpandey) [inline Tests/RudderStackAnalyticsTests/StateManagementTests/ResetUserIdentityActionTests.swift:1]: Please refer to the Kotlin PR, or [here](https://github.com/rudderlabs/rudder-sdk-kotlin/blob/3c8e230a749b376bcff3952f448096918877a23b/core/src/test/kotlin/com/rudderstack/sdk/kotlin/core/internals/models/useridentity/ResetUserIdentityActionTest.kt) for reference. 

I don't see any new tests added related to the code changes. Please add new unit tests to cover the new code changes.

**Review by 1abhishekpandey — COMMENTED** [review]
Nice separation of work.

In the PR description, `Additional Context` section, please add the next items you're planning to work, related to this RESET feature. This will give clear visibility to the reviewer.

For the public facing API's and classes, let's ensure we have added the proper documentation. This will help the user of the API to better understand the API.

File: Sources/RudderStackAnalytics/StateManagement/ResetOptions.swift (hunk 2 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/StateManagement/ResetOptions.swift]: Currently, no new items are planned for this class, so keeping `ResetEntries` inside `ResetOptions` is fine. If more items are added later, we can move it to a separate file.

Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/StateManagement/ResetOptions.swift]: I understand it's fine, but let's try to separate the things now while we are developing the feature to avoid future tech debt!

File: Tests/RudderStackAnalyticsTests/RudderStackAnalyticsTests.swift (hunk 7 above)
Comment (SKannaniOS) [reply Tests/RudderStackAnalyticsTests/RudderStackAnalyticsTests.swift]: Test case reversed.

File: Tests/RudderStackAnalyticsTests/StateManagementTests/ResetUserIdentityActionTests.swift
Comment (SKannaniOS) [reply Tests/RudderStackAnalyticsTests/StateManagementTests/ResetUserIdentityActionTests.swift:1]: New test cases added.

File: Sources/RudderStackAnalytics/Base/Analytics.swift (hunk 4 above)
Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/Base/Analytics.swift:211]: Have it like this: (Taken from Kotlin - current description is a one liner and it doesn't seems to provide enough context.)

/**
     * Resets the user identity to its initial state with iOS-specific functionality.
     *
     * By default, this method:
     * - Generates a new anonymous ID (clears the existing one)
     * - Clears the user ID (sets it to empty string)
     * - Clears user traits (resets to empty JSON object)
     * - Refreshes the user session (Android-specific behaviour)
     *
     * Parameter options: iOS-specific [ResetOptions] to override the default reset behaviour.
     *                When provided, the `ResetEntries` configuration within these options
     *                allows selective control over which data entries are reset:
     *                - `anonymousId`: Controls whether to generate a new anonymous ID
     *                - `userId`: Controls whether to clear the user ID
     *                - `traits`: Controls whether to clear user traits
     *                - `session`: Controls whether to refresh the session (Android-specific)
     *                These options override the default behaviour of resetting all user data.
     */

**Review by 1abhishekpandey — APPROVED** [review]

**Review by vgupta98 — APPROVED** [review]
